worker/worker/parser/category.py

```diff
@@ -16,97 +18,152 @@ def extract_challenge_ids(txt):
     return result
 
 
-def extract_categories(txt):
-    pattern = '<li><a class="submenu.*?" href="fr/Challenges/(.*?)/" '
-    result = re.findall(pattern, txt)
+def extract_categories(r):
+    tree = html.fromstring(r.content)
+    result = tree.xpath('//li/a[starts-with(@class, "submenu")][starts-with(@href, "fr/Challenges")]/@href')
+    result = list(map(lambda x: x.split('/')[2], result))
     if not result:
         raise RootMeParsingError("Could not parse categories.")
     return result
 
 
-def extract_category_logo(txt):
-    txt = txt.replace('\n', '')
-    txt = txt.replace('&nbsp;', '')
-
-    pattern = '<img class=\'vmiddle\' alt="" src="(.*?)" .*?/>'
-    result = re.findall(pattern, txt)
+def extract_category_logo(r):
+    tree = html.fromstring(r.content)
+    result = tree.xpath('//h1/img[@class="vmiddle"][starts-with(@src, "local")]/@src')
     if not result or not result[0]:
```

Comment:
`not result[0]:` semble etre redondant non?

Si result est None `if result` va pas passer: OK
Si result est [] `if result` va pas passer: OK
Si result a 1 ou plus elements ca va passer: OK

du coup not result[0]: est appelé quand? Tu peux avoir `[None]` en retour de tree.xpath?

Reply:
Tu peux avoir `['']` dans certain cas (en génaral si tu finis ton xpath par un '/text()'

Reply:
Je vais adapter suivant les cas et préciser en commentaire